agent/auto_config.go

```diff
@@ -0,0 +1,263 @@
+package agent
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/consul/agent/agentpb"
+	"github.com/hashicorp/consul/agent/config"
+	"github.com/hashicorp/consul/lib"
+)
+
+const (
+	// autoConfigFileName is the name of the file that the agent auto-config settings are
+	// stored in within the data directory
+	autoConfigFileName = "auto-config.json"
+)
+
+// loadConfig will build the agents configuration including the extraHead source injected
+// after all other defaults but before any user supplied configuration.
+func loadConfig(flags config.Flags, extraHead config.Source) (*config.RuntimeConfig, []string, error) {
+	b, err := config.NewBuilder(flags)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if extraHead.Data != "" {
+		b.Head = append(b.Head, extraHead)
+	}
+
+	cfg, err := b.BuildAndValidate()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return &cfg, b.Warnings, nil
+}
+
+// ReadConfig is used to reparse the configuration taking into account the agent
+// auto-config settings
+func (a *Agent) ReadConfig() (*config.RuntimeConfig, []string, error) {
+	a.stateLock.Lock()
+	defer a.stateLock.Unlock()
```

Comment:
Why does state need to be locked here? Would be good capture that in the docstring.

Reply:
This method is removed. It was because we could be reloading the config at the same time as populating the autoConfigJSON value and it was to prevent that racing. The code changed a bit so this function is gone.